Sources/PackageCollections/API.swift

```diff
@@ -248,3 +251,192 @@ public enum PackageCollectionError: Equatable, Error {
 
     case unsupportedPlatform
 }
+
+// MARK: - Package index
+
+public protocol PackageIndexProtocol {
+    /// Sets package index URL.
+    ///
+    /// - Parameters:
+    ///   - callback: The closure to invoke when result becomes available
+    func set(
+        url: Foundation.URL,
+        callback: @escaping (Result<Void, Error>) -> Void
+    )
+
+    /// Unsets package index URL.
+    ///
+    /// - Parameters:
+    ///   - callback: The closure to invoke when result becomes available
+    func unset(callback: @escaping (Result<Void, Error>) -> Void)
+    
+    /// Returns package index URL.
+    ///
+    /// - Parameters:
+    ///   - callback: The closure to invoke when result becomes available
+    func get(callback: @escaping (Result<Foundation.URL?, Error>) -> Void)
+
+    /// Returns metadata for the package identified by the given `PackageIdentity`.
+    ///
+    /// A failure is returned if the package is not found.
+    ///
+    /// - Parameters:
+    ///   - identity: The package identity
+    ///   - location: The package location (optional for deduplication)
+    ///   - callback: The closure to invoke when result becomes available
+    func getPackageMetadata(
+        identity: PackageIdentity,
+        location: String?,
+        callback: @escaping (Result<PackageCollectionsModel.PackageMetadata, Error>) -> Void
+    )
+
+    /// Finds and returns packages that match the query.
+    ///
+    /// - Parameters:
+    ///   - query: The search query
+    ///   - callback: The closure to invoke when result becomes available
+    func findPackages(
+        _ query: String,
+        callback: @escaping (Result<PackageCollectionsModel.PackageSearchResult, Error>) -> Void
+    )
+    
+    /// A paginated list of packages in the index.
+    ///
+    /// - Parameters:
+    ///   - offset: Offset of the first item in the result
+    ///   - limit: Number of items to return in the result. Implementations might impose a threshold for this.
+    ///   - callback: The closure to invoke when result becomes available
+    func listPackages(
+        offset: Int,
+        limit: Int,
+        callback: @escaping (Result<PackageCollectionsModel.PaginatedPackageList, Error>) -> Void
+    )
+}
+
+public enum PackageIndexError: Equatable, Error {
+    /// Package index support is disabled
+    case featureDisabled
+    /// No package index configured
+    case notConfigured
+}
+
+// MARK: - Package index and collection combined
+
+// It's on purpose that the protocol does not conform to `PackageCollectionsProtocol` and `PackageIndexProtocol`,
+// so that we can exclude APIs that we believe are not required and have more flexibility around return types.
```

Comment:
👍 

unless we need it for mocking / testing, maybe we dont even need a dedicated protocol and the concrete impl is enough

Reply:
removed https://github.com/apple/swift-package-manager/pull/4010/commits/c1d364a022d867b5bfcd9f69006a1fc714001414